src/metabase/upload.clj

```diff
@@ -170,18 +173,57 @@
     "unnamed_column"
     (u/slugify (str/trim raw-name))))
 
+(defn- is-pk?
+  [[col-name type]]
+  (and (#{"id" "pk"} (u/lower-case-en (name col-name)))
+       (isa? type ::int)))
+
+(defn- add-pk-column
+  "Adds driver-specific auto-increment and non-nil/PK constraints to the first ID/PK column it finds; otherwise prepends
+  a new ID column to the columns list."
+  [driver name-type-pairs]
+  (let [found-it   (atom false)
+        pked-pairs (doall (map (fn [[name _type :as p]]
+                                 (if (and (not @found-it)
+                                          (is-pk? p))
+                                   (do (reset! found-it true)
+                                       [(keyword name) {:type ::pk :opts (driver/pk-options driver)}])
+                                   p))
+                               name-type-pairs))]
+    (if @found-it ;; we already have a PK, just return them
+      pked-pairs
+      ;;otherwise, prepend a new ID column
+      (cons [:id {:type ::pk :opts (driver/pk-options driver) :exclude-in-insert? true}]
+            name-type-pairs))))
+
+(defn- upload-type->column-spec
+  "Converts our internal types (e.g., ::int) to the type expected by the database as a HoneySQL
+  keyword (e.g., :biginteger). However, the returned type is *always* a vector since there may be additional HoneySQL
+  options included. Sample return values:
+
+    - [:text]
+    - [[:varchar 255]]
+    - [:int :autoincrement [:not nil]]"
+  [driver type-or-type-map]
+  (if (map? type-or-type-map)
+    (let [{the-type :type opts :opts} type-or-type-map]
+      (vec (cons (driver/upload-type->database-type driver the-type) opts)))
```

Comment:
I find `into` more idiomatic than `(vec (cons ...)` in clojure because it's faster. Feel free to ignore this though, this is bikeshedding to the max
```
(time (do (into [1] (range 2 100000)) nil))
"Elapsed time: 2.012625 msecs"

(time (do (vec (cons 1 (range 2 100000))) nil))
"Elapsed time: 11.779375 msecs"
```